### PR TITLE
Reformatted BTD fields

### PR DESCRIPTION
**Pull request by RevathiJambunathan**
In this PR, back-transformed diagnostics is reformatted to use the new diagnostics framework.
The field buffer data is flushed out using the customized format from the existing BTD. These functions in the reformatted BTD are named with a prefix `TMP_`, since this is a temporary fix, until support for plotfile/OpenPMD format is added. (Issue #1167 )

An automated test for comparing the sliced-BTD with the full BTD snapshot is included for the reformatted BTD as WarpX-tests.ini titled "Reformatted_BTD_ReducedSliceDiag".

Coarsening ratio of (1 1 1) is only supported at this time, since CoarsenIO:: does not fill ghost cells of the destination cell-centered MultiFab. This ghost data is required when averaging-down the field-data to a single-level for BTD. 
This will be fixed in CoarsenIO in a follow-up PR and then the limitation on coarsening ratio will be removed. 

Only support for fields is included. Particle-buffers and their Lorentz-transform will be added in a future PR. (Issue #1166 )

The input parameters to be added to use the new BTD is as follows
```

diagnostics.diags_names = diag1 diag2
diag1.period = 10
diag1.diag_type = Full
diag2.diag_type = BackTransformed
diag2.do_back_transformed_fields = 1
diag2.num_snapshots_lab = 2
diag2.dt_snapshots_lab = 1.8679589331096515e-13
diag2.fields_to_plot = Ex Ey Ez Bx By Bz jx jy jz rho
```
Support for any user-defined `diag_lo` and `diag_hi`is also included.

I have verified the implementation by performing a 2D test-case in Examples/Modules/RigidInjection 
The test for back-transformed slice or line data is also performed and verified for accuracy.
Screenshot from 2020-07-08 21-19-34

Input file for the RigidInjection test:

[inputs_2d_BoostedFrame_reformatted.txt](https://github.com/ECP-WarpX/WarpX/files/4949975/inputs_2d_BoostedFrame_reformatted.txt)

Analysis script (renamed the extension of this .py script to .txt since github dint allow upload of file with .py.)
[verify_reformatted_BTD.txt](https://github.com/ECP-WarpX/WarpX/files/4949991/verify_reformatted_BTD.txt)

### CODE REVIEW

File: Examples/Modules/boosted_diags/inputs_3d_slice_reformatted
Comment (atmyers) [inline Examples/Modules/boosted_diags/inputs_3d_slice_reformatted]: Do we need both this and `inputs_3d_slice`?

File: Source/Diagnostics/BTDiagnostics.H
Comment (atmyers) [inline Source/Diagnostics/BTDiagnostics.H]: spelling : temporary

File: Source/Evolve/WarpXEvolve.cpp
Comment (atmyers) [inline Source/Evolve/WarpXEvolve.cpp]: Do we still need this?

File: Source/Diagnostics/BTDiagnostics.H
Comment (RevathiJambunathan) [reply Source/Diagnostics/BTDiagnostics.H]: Thank you for catching this. I fixed it now.

File: Source/Evolve/WarpXEvolve.cpp
Comment (RevathiJambunathan) [reply Source/Evolve/WarpXEvolve.cpp]: I kept this one to be consistent with development branch and removed the previous one since it was added to compare the old and new BTD implementation.

File: Examples/Modules/boosted_diags/inputs_3d_slice_reformatted
Comment (RevathiJambunathan) [reply Examples/Modules/boosted_diags/inputs_3d_slice_reformatted]: We could remove the old one actually. The old one tests the old implementation and new tests the new implementation. Will do whatever you suggest :)

Comment (atmyers) [reply Examples/Modules/boosted_diags/inputs_3d_slice_reformatted]: I don't think it makes a big difference but IMO there's slightly less chance of confusion if there's only one inputs file.

File: Regression/WarpX-tests.ini
Comment (atmyers) [inline Regression/WarpX-tests.ini]: If we turn this test off, maybe we should create an issue to keep track of the fact that we need to further split up the tests into more CI runners.

**Comment by atmyers** [comment]
Looks like there is a conflict to be fixed.

**Comment by RevathiJambunathan** [comment]
Thank you for reviewing @atmyers .
I just update the branch with development and fixed the conflict. Also addressed some unused parameter warnings from one of the Diagnostic functions added in this PR, in c33e40c

**Comment by RevathiJambunathan** [comment]
I have removed the CI-test for the reformatted BTD slices and opened Issue #1234 to ensure the test is added when the time-out issue is resolved. Thank you for the suggestion @atmyers

**Review by atmyers — APPROVED** [review]

File: Source/Diagnostics/BTDiagnostics.H
Comment (MaxThevenet) [inline Source/Diagnostics/BTDiagnostics.H]:
```suggestion
    /** \brief Flush m_mf_output and particles to file.
```

File: Source/Diagnostics/BTDiagnostics.H
Comment (MaxThevenet) [inline Source/Diagnostics/BTDiagnostics.H]: I think it's alright to put this in the Doxygen doc, but that's just a suggestion
```suggestion
     * Currently, a temporary customized output format for the buffer
     * data is implemented and called in this function.
     * After implementing OpenPMD and plotfile formats, this function
     * can be defined and implemented in Diagnostics.H, such that,
     * the function call to flush out buffer data for
     * FullDiagnostics and BTDiagnostics is the same */
```

File: Source/Diagnostics/BTDiagnostics.H
Comment (MaxThevenet) [inline Source/Diagnostics/BTDiagnostics.H]: Do I understand this well?
```suggestion
    /** This function allocates and initializes particle buffers for all the snapshots.
     * This is currently an empty function:
     * The particle containers required for this must be added to populate this function.
     */
```

File: Source/Diagnostics/BTDiagnostics.H
Comment (MaxThevenet) [inline Source/Diagnostics/BTDiagnostics.H:39]: Would sth like that work?
```suggestion
    bool DoComputeAndPack (int step, bool force_flush=false) {return step>=0;} override;
```

File: Source/Diagnostics/BTDiagnostics.cpp
Comment (MaxThevenet) [inline Source/Diagnostics/BTDiagnostics.cpp]: If the short version in the .H file works, this could be removed.

**Review by MaxThevenet — APPROVED** [review]
Thanks for this PR @RevathiJambunathan! Please find minor comments in the text.

After this PR is merged, do you think this is work adding a short description in the dev documentation at https://warpx.readthedocs.io/en/latest/developers/diagnostics.html#back-transformed-diagnostics?

This doesn't have to be long, but the main steps could be (there may be mistakes below, feel free to correct):
- field buffers with fixed size are allocated at init
- At each iteration, the CC data for all fields is computed by `BTDiagnostics:: m_all_field_functors` and stored in `BTDiagnostics::m_cell_centered_data`, and the appropriate slice is Lorentz-transformed and appended to the buffer by ???
- When a buffer is full, it is flushed to disk with ???

Thanks!

File: Source/Diagnostics/BTDiagnostics.H
Comment (RevathiJambunathan) [reply Source/Diagnostics/BTDiagnostics.H]: YEs. it is correct that the function is currently empty and particle containers are required to populate this function

File: Source/Diagnostics/BTDiagnostics.cpp
Comment (RevathiJambunathan) [reply Source/Diagnostics/BTDiagnostics.cpp]: It was leading to compiler error. I reverted the implementation to cpp, but with return(step>=0)

**Comment by RevathiJambunathan** [comment]
Thank you @MaxThevenet for reviewing the PR and for the suggestions.

**Review by MaxThevenet — APPROVED** [review]

**Comment by RevathiJambunathan** [comment]
@MaxThevenet to ensure that the suggested addition to documentation is made, I have included it in the re-organize diagnostics project in the "to-do" column - so I can keep track

**Comment by MaxThevenet** [comment]
Thanks! I'll merge when tests pass.

**Comment by MaxThevenet** [comment]
:tada:

**Pull request merged**
